### PR TITLE
Upgrade GLFW dependency

### PR DESCRIPTION
**Pull request by archbirdplus**
- [x] Declare GLFW sources as a dependency in build.zig.zon.
- [x] Update GLFW to the very latest.
- [x] Remove GLFW headers.
- [x] Patch glad.c for X11 on Mac so it actually works.

This PR is based on top of the [previous](https://github.com/PixelGuys/Cubyz-libs/pull/1), meaning it also will:
- [x] Add KHR headers.
- [x] Add ALSA headers.
- [x] Create Zig Build steps for each/all target `zig build install release`.

It will NOT automatically tar.gz the artifacts.

The GLFW package looks like this:
```
cubyz-sources
├── include
│   └── GLFW
│       ├── glfw3.h
│       └── glfw3native.h
└── src
    ├── glx_context.c
    └── ...
```
`build.zig` is built for the latest version of GLFW (origin/master), but if you want Mac support you need [my fork](https://github.com/archbirdplus/glfw/tree/mac-x11). You still cannot cross-compile for Mac.

### What's left:
* Check that the new stuff compiles on your/other machines (especially due to ALSA's `version.h`).
* Build/release these artifacts.
* Fetch these artifacts from Cubyz.

### CODE REVIEW

File: build.zig [hunk 1]
```diff
@@ -132,38 +135,47 @@ pub fn addFreetypeAndHarfbuzz(b: *std.Build, c_lib: *std.Build.Step.Compile, tar
...
+	const win32 = tag == .windows;
+	const linux = tag == .linux;
+	const macos = tag == .macos;
+	// in the future, there might be another Mac option besides x11
+	const x11 = linux or macos;
+	const cocoa = !x11 and !win32;
+	const ws_flag = if(x11) "-D_GLFW_X11" else
+					if(win32) "-D_GLFW_WIN32" else
+					if(cocoa) "-D_GLFW_COCOA" else
+					unreachable;
```
Comment (IntegratedQuantum) [inline build.zig]: I'm not a fan of line expansions like this. I would prefer something like
```zig
	const ws_flag = if(x11) "-D_GLFW_X11"
	else if(win32) "-D_GLFW_WIN32"
	else if(cocoa) "-D_GLFW_COCOA"
	else unreachable;
```
or
```zig
	const ws_flag = if(x11)
		"-D_GLFW_X11"
	else if(win32)
		"-D_GLFW_WIN32"
	else if(cocoa)
		"-D_GLFW_COCOA"
	else
		unreachable;
```

File: build.zig [hunk 2]
```diff
@@ -132,38 +135,47 @@ pub fn addFreetypeAndHarfbuzz(b: *std.Build, c_lib: *std.Build.Step.Compile, tar
...
+	const glfw = b.dependency("glfw", .{});
+	const root = glfw.path("src");
+	const tag = target.result.os.tag;
+
+	const win32 = tag == .windows;
+	const linux = tag == .linux;
+	const macos = tag == .macos;
+	// in the future, there might be another Mac option besides x11
+	const x11 = linux or macos;
+	const cocoa = !x11 and !win32;
```
Comment (IntegratedQuantum) [inline build.zig]: There is also wayland on linux which might make sense to support in the future:
```
const cocoa = !x11 and !win32 and macos;
const wayland = !x11 and !win32 and linux;
```

File: build.zig [hunk 3]
```diff
@@ -106,6 +108,7 @@ pub inline fn addPortAudio(b: *std.Build, c_lib: *std.Build.Step.Compile, target
 }
 
 pub fn addFreetypeAndHarfbuzz(b: *std.Build, c_lib: *std.Build.Step.Compile, target: std.Build.ResolvedTarget, optimize: std.builtin.OptimizeMode, flags: []const []const u8) void {
+	// TODO: delete?
```
Comment (IntegratedQuantum) [inline build.zig]: What exactly do you want to delete here?

File: build.zig.zon [hunk 4]
```diff
@@ -3,6 +3,9 @@
 	.version = "0.0.0",
 	.paths = .{""},
 	.dependencies = .{
+		.glfw = .{
+			.path = "../glfw/cubyz-sources"
+		},
```
Comment (IntegratedQuantum) [inline build.zig.zon:9]: You should get glfw from their official github directly.
Checking out glfw locally should not be a requirement to build Cubyz.
`.url = "https://github.com/glfw/glfw/archive/refs/tags/3.4.tar.gz"` for the latest release.

**Review by IntegratedQuantum — CHANGES_REQUESTED** [review]
See code comments.

Additionally it might be worth thinking about using enums instead of boolean flags:
```zig
const Backend = enum{.win32, .x11, .cocoa};
...
const backend: Backend = if(tag == windows) .win32
else if(tag == .linux or tag == .macos) .x11
else std.log.err("Unsupported target OS: {}", .{tag});

const ws_flag = switch(backend) {
	.win32 => "-D_GLFW_WIN32",
	.x11 => "-D_GLFW_X11",
	.cocoa => "-D_GLFW_COCOA",
};
...
.{.condition = backend == .win32, .files = ...}
...
```

With that it would also be possible to simplify the `fileses` code to something like this:
```zig
const fileses: [3][]const []const u8 = .{
	&.{base files},
	switch(tag) {
		.linux => &.{linux files},
		.macos => &.{mac files},
		.else => &.{},
	},
	switch(backend) {
		.win32 => &.{win32 files},
		.x11 => &.{x11 files},
		.cocoa => &.{cocoa files},
	},
};
for(fileses) |files| {
	c_lib.addCSourceFiles(.{.root = root, .files = files, .flags = all_flags.items});
}
```

File: build.zig.zon (hunk 4 above)
Comment (archbirdplus) [reply build.zig.zon:9]: Sure. Especially until at least Mac can cross-compile.

File: build.zig (hunk 2 above)
Comment (archbirdplus) [reply build.zig]: Added as a TODO comment. But I consider this out of scope for this PR.

File: build.zig (hunk 1 above)
Comment (archbirdplus) [reply build.zig]: I just replaced it with enums. This may be more fragile when extending the builds to Wayland, but we'll see.

**Review by IntegratedQuantum — APPROVED** [review]
It seems to work, thank you!

**Pull request merged**